### PR TITLE
[xray] Implementing Gcs sharding

### PR DESCRIPTION
**Pull request by heyucongtom**
## What do these changes do?

Basically a re-implementation of #2281, with modifications of #2298 (A fix of #2334, for rebasing issues.).
[+] Implement sharding for gcs tables.
[+] Keep ClientTable and ErrorTable managed by the primary_shard. TaskTable is managed by the primary_shard for now, until a good hashing for tasks is implemented.
[+] Move AsyncGcsClient's initialization into Connect function.
[-] Move GetRedisShard and bool sharding from RedisContext's connect into AsyncGcsClient. This may make the interface cleaner.

### CODE REVIEW

**Comment by heyucongtom** [comment]
@robertnishihara 
"bash ../src/plasma/test/run_tests.sh" this test seems to be failing. 
Also I need some help to check the sharding scheme for request-notification & subscribeAsync. Still on the way of reading through the codebase.

**Comment by robertnishihara** [comment]
In this PR, you should get rid of this if statement

https://github.com/ray-project/ray/blob/2a3b02649a799ed0bcfa4b98e0bbb5a3bcea4f3c/test/runtest.py#L1157-L1159

so that the sharded case will be tested more.

File: src/ray/gcs/client.cc [hunk 1]
```diff
@@ -37,8 +104,41 @@ AsyncGcsClient::AsyncGcsClient(CommandType command_type)
 AsyncGcsClient::AsyncGcsClient() : AsyncGcsClient(ClientID::from_random()) {}
 
 Status AsyncGcsClient::Connect(const std::string &address, int port, bool sharding) {
-  RAY_RETURN_NOT_OK(context_->Connect(address, port, sharding));
-  RAY_RETURN_NOT_OK(primary_context_->Connect(address, port, /*sharding=*/false));
+  RAY_RETURN_NOT_OK(primary_context_->Connect(address, port));
+  std::vector<std::string> addresses{address};
+  std::vector<int> ports{port};
```
Comment (robertnishihara) [inline src/ray/gcs/client.cc]: Why do we initialize `addresses` and `ports` with the `address` and `port` of the primary shard? Don't we want them to start out empty and then contain just the shard addresses/ports?

File: src/ray/gcs/redis_context.h [hunk 2]
```diff
@@ -51,7 +51,7 @@ class RedisContext {
   RedisContext()
       : context_(nullptr), async_context_(nullptr), subscribe_context_(nullptr) {}
   ~RedisContext();
-  Status Connect(const std::string &address, int port, bool sharding);
+  Status Connect(const std::string &address, int port);
```
Comment (robertnishihara) [inline src/ray/gcs/redis_context.h]: @pcmoritz added this `sharding` flag just for the tests, so that we could run some of the tests with a single Redis shard and not always need to start up at least 2 shards.

This shouldn't have an effect on regular execution, but we may want to add it back for the purpose of simplifying some tests.

File: src/ray/gcs/tables.h [hunk 3]
```diff
@@ -175,9 +175,23 @@ class Log : virtual public PubsubInterface<ID> {
   Status CancelNotifications(const JobID &job_id, const ID &id,
                              const ClientID &client_id);
 
+  void AddShards(std::vector<std::shared_ptr<RedisContext>> &contexts) {
+    if (shard_contexts_.size() > 0) {
+      return;  // Only add shards if at the first time.
```
Comment (robertnishihara) [inline src/ray/gcs/tables.h]: I'm not sure if we want to keep this method, it seems to be a sign that we're initializing things in the wrong order (which we probably are), but anyway, this should be `RAY_CHECK(shard_contexts_.size() == 0)` instead of the if statement I believe.

File: src/ray/gcs/tables.h [hunk 4]
```diff
@@ -175,9 +175,23 @@ class Log : virtual public PubsubInterface<ID> {
   Status CancelNotifications(const JobID &job_id, const ID &id,
                              const ClientID &client_id);
 
+  void AddShards(std::vector<std::shared_ptr<RedisContext>> &contexts) {
```
Comment (robertnishihara) [inline src/ray/gcs/tables.h]: I'd like to get rid of this method. However, it may be easier to tackle it in a separate PR.

That would mean that we need to pass in the `contexts` into the constructor of this class. Which means we would need to connect to the primary Redis shard (to get the number of shards) before calling the table/log constructor, which means we would need to connect to the primary redis shard in the gcs client constructor, which is probably reasonable, but may cause other issues.

@stephanie-wang any thoughts about this?

**Review by robertnishihara — COMMENTED** [review]
@heyucongtom this looks quite good, nice work! Is it currently working? I left some comments/questions.

File: src/ray/gcs/tables.h (hunk 4 above)
Comment (heyucongtom) [reply src/ray/gcs/tables.h]: I guess this is one of the most important point causing some suspicious failures.

File: src/ray/gcs/client.cc [hunk 5]
```diff
@@ -2,22 +2,89 @@
...
-  client_table_.reset(new ClientTable(primary_context_, this, client_id));
-  object_table_.reset(new ObjectTable(context_, this, command_type));
-  actor_table_.reset(new ActorTable(context_, this));
-  task_table_.reset(new TaskTable(context_, this, command_type));
-  raylet_task_table_.reset(new raylet::TaskTable(context_, this, command_type));
-  task_reconstruction_log_.reset(new TaskReconstructionLog(context_, this));
-  heartbeat_table_.reset(new HeartbeatTable(context_, this));
-  error_table_.reset(new ErrorTable(primary_context_, this));
-  profile_table_.reset(new ProfileTable(context_, this));
+  std::vector<std::shared_ptr<RedisContext>> tmp_primary;
```
[the remark this answers is not in the document]
Comment (heyucongtom) [reply src/ray/gcs/client.cc]: Done.

File: src/ray/gcs/redis_context.h (hunk 2 above)
Comment (heyucongtom) [reply src/ray/gcs/redis_context.h]: Done

File: src/ray/gcs/tables.h [hunk 6]
```diff
@@ -175,9 +175,23 @@ class Log : virtual public PubsubInterface<ID> {
...
+      return;  // Only add shards if at the first time.
+    }
+    for (std::shared_ptr<RedisContext> context : contexts) {
+      shard_contexts_.push_back(context);
+    }
+  }
+
  protected:
+  template <typename AnyID>
+  std::shared_ptr<RedisContext> GetRedisContext(AnyID &&id) {
```
[the remark this answers is not in the document]
Comment (heyucongtom) [reply src/ray/gcs/tables.h]: Yep. Done.

File: src/ray/gcs/client.cc (hunk 1 above)
Comment (heyucongtom) [reply src/ray/gcs/client.cc]: Yep. I mixed up the "no-sharding scenario", where all of those share the primary shard, and "sharding scenario", where they use separate shards here.

**Comment by robertnishihara** [comment]
@heyucongtom this looks really good! If you run

```
RAY_USE_XRAY=1 python test/runtest.py APITestSharded
```

Does that run successfully?

**Comment by robertnishihara** [comment]
@heyucongtom the existing tests, in particular `RAY_USE_XRAY=1 python test/runtest.py APITestSharded` should be sufficient, so I don't think we'll need to add any new tests.

Now that https://github.com/ray-project/ray/pull/2289 has been merged, it may be possible to re-enable this test https://github.com/ray-project/ray/blob/8ae82180b4e1e5eddfb44ff8b96837fdd539d869/test/monitor_test.py#L110-L118

by getting rid of the `os.environ.get("RAY_USE_XRAY") == "1"` skip statement.

File: src/ray/gcs/client.cc [hunk 7]
```diff
@@ -2,44 +2,135 @@
 
 #include "ray/gcs/redis_context.h"
 
+static void GetRedisShards(redisContext *context, std::vector<std::string> *addresses,
+                           std::vector<int> *ports) {
```
Comment (stephanie-wang) [inline src/ray/gcs/client.cc]: Can we change these pointers to references?

File: src/ray/gcs/client.cc [hunk 8]
```diff
@@ -2,44 +2,135 @@
...
-  task_reconstruction_log_.reset(new TaskReconstructionLog(context_, this));
-  task_lease_table_.reset(new TaskLeaseTable(context_, this));
-  heartbeat_table_.reset(new HeartbeatTable(context_, this));
-  error_table_.reset(new ErrorTable(primary_context_, this));
-  profile_table_.reset(new ProfileTable(context_, this));
+  // A boolean placeholder value.
+  // Used since changing the API of Connect is breaking some tests.
+  // Would be removed soon in the future.
+  const bool DUMMY_BOOL = true;
+  RAY_CHECK_OK(primary_context_->Connect(address, port, DUMMY_BOOL));
```
Comment (stephanie-wang) [inline src/ray/gcs/client.cc]: Hmm, does it make sense to have a default value instead of the `DUMMY_BOOL` in the `Connect` method? I'm not clear on what the value is for, though, so I'm not sure if that's the way to fix this.

How hard is it to remove the argument completely as part of this PR?

File: src/ray/gcs/client.cc [hunk 9]
```diff
@@ -2,44 +2,135 @@
...
+  std::vector<std::string> addresses;
+  std::vector<int> ports;
+  GetRedisShards(primary_context_->sync_context(), &addresses, &ports);
+  if (addresses.size() == 0 || ports.size() == 0) {
+    addresses.push_back(address);
+    ports.push_back(port);
+  }
+
+  // First time connect: populate shard_contexts.
+  if (shard_contexts_.empty()) {
```
Comment (stephanie-wang) [inline src/ray/gcs/client.cc]: This should always be true, right?

File: src/ray/gcs/client.cc [hunk 10]
```diff
@@ -2,44 +2,135 @@
...
 
-AsyncGcsClient::AsyncGcsClient(CommandType command_type)
-    : AsyncGcsClient(ClientID::from_random(), command_type) {}
+AsyncGcsClient::AsyncGcsClient(const std::string &address, int port, CommandType command_type)
+    : AsyncGcsClient(address, port, ClientID::from_random(), command_type) {}
 
-AsyncGcsClient::AsyncGcsClient() : AsyncGcsClient(ClientID::from_random()) {}
+AsyncGcsClient::AsyncGcsClient(const std::string &address, int port) : AsyncGcsClient(address, port, ClientID::from_random()) {}
 
 Status AsyncGcsClient::Connect(const std::string &address, int port, bool sharding) {
```
Comment (stephanie-wang) [inline src/ray/gcs/client.cc]: Is it possible to remove this method now?

File: src/ray/gcs/client.cc [hunk 11]
```diff
@@ -54,9 +145,15 @@ Status Attach(plasma::EventLoop &event_loop) {
 }
 
 Status AsyncGcsClient::Attach(boost::asio::io_service &io_service) {
-  asio_async_client_.reset(new RedisAsioClient(io_service, context_->async_context()));
-  asio_subscribe_client_.reset(
-      new RedisAsioClient(io_service, context_->subscribe_context()));
+  // Take care of sharding contexts.
+  shard_asio_async_clients_.clear();
+  shard_asio_subscribe_clients_.clear();
```
Comment (stephanie-wang) [inline src/ray/gcs/client.cc]: Why are the `clear()` calls necessary? Do we expect them to have something in them?

I would lean towards making sure `Attach` is only called once, and I would change this to a `RAY_CHECK(shard_asio_async_clients_.empty())`.

File: src/ray/gcs/tables.h [hunk 12]
```diff
@@ -481,10 +497,11 @@ class ErrorTable : private Log<JobID, ErrorTableData> {
                            const std::string &error_message, double timestamp);
 };
 
-class ProfileTable : private Log<UniqueID, ProfileTableData> {
+class ProfileTable : public Log<UniqueID, ProfileTableData> {
```
Comment (stephanie-wang) [inline src/ray/gcs/tables.h]: Why the change to `public`?

File: test/runtest.py [hunk 13]
```diff
@@ -1215,11 +1215,7 @@ def init_ray(self, **kwargs):
         if kwargs is None:
             kwargs = {}
         kwargs["start_ray_local"] = True
-        if os.environ.get("RAY_USE_XRAY") == "1":
-            print("XRAY currently supports only a single Redis shard.")
-            kwargs["num_redis_shards"] = 1
-        else:
-            kwargs["num_redis_shards"] = 20
+        kwargs["num_redis_shards"] = 20
```
Comment (stephanie-wang) [inline test/runtest.py]: Awesome :)

**Review by stephanie-wang — CHANGES_REQUESTED** [review]
Thanks, this looks great! I left a few comments.

File: test/runtest.py (hunk 13 above)
Comment (heyucongtom) [reply test/runtest.py]: 💯

File: src/ray/gcs/client.cc (hunk 7 above)
Comment (heyucongtom) [reply src/ray/gcs/client.cc]: Done

File: src/ray/gcs/client.cc (hunk 8 above)
Comment (heyucongtom) [reply src/ray/gcs/client.cc]: @robertnishihara suggests that we keep the redis_context::Connect(address, port, sharding) interface intact for testing purpose. I suggest using another PR to fix this.

File: src/ray/gcs/client.cc (hunk 9 above)
Comment (heyucongtom) [reply src/ray/gcs/client.cc]: In the case that it is connected for only once. I think this shall be true, and probably I shall add an assert-style statement here.

File: src/ray/gcs/client.cc (hunk 10 above)
Comment (heyucongtom) [reply src/ray/gcs/client.cc]: There is a //TODO here which may need us to call client table's connect. I moved it up to the ctor

File: src/ray/gcs/client.cc (hunk 11 above)
Comment (heyucongtom) [reply src/ray/gcs/client.cc]: Sounds good :)

File: src/ray/gcs/tables.h (hunk 12 above)
Comment (heyucongtom) [reply src/ray/gcs/tables.h]: Originally we want to use the AddShards function. Change reverted.

File: src/ray/gcs/client.cc [hunk 14]
```diff
@@ -2,8 +2,8 @@
 
 #include "ray/gcs/redis_context.h"
 
-static void GetRedisShards(redisContext *context, std::vector<std::string> *addresses,
-                           std::vector<int> *ports) {
+static void GetRedisShards(redisContext *context, std::vector<std::string>& addresses,
+                           std::vector<int>& ports) {
```
Comment (stephanie-wang) [inline src/ray/gcs/client.cc]: Sorry for the nit, but to keep the style consistent, we usually put the `&` with the variable name, so `&addresses` and `&ports` instead of `& addresses` and `& ports`.

**Review by stephanie-wang — APPROVED** [review]
Thanks! Left one small comment, but otherwise looks great :)

**Comment by heyucongtom** [comment]
Thanks! I would wait for the tests to finish and try linting it a little
bit.

**Comment by heyucongtom** [comment]
@robertnishihara @stephanie-wang There's some error where "GET NumRedisShards" is getting a nil reply after I move the things into the constructor. My guess is that this command is not properly setup when the constructor is called. Please feel free to take a look at it 

/home/travis/build/ray-project/ray/src/ray/gcs/client.cc:23 Check failed: num_attempts < RayConfig::instance().redis_db_connect_retries() No entry found for NumRedisShards

File: src/ray/gcs/client.cc (hunk 8 above)
Comment (robertnishihara) [reply src/ray/gcs/client.cc]: How about we just pass in `true` instead of defining `DUMMY_BOOL`. Maybe something like pass in `true /* sharding */`.

I agree it would be best to just have a default argument, but it's ok to do that separately.

File: src/ray/gcs/client.cc [hunk 15]
```diff
@@ -2,51 +2,150 @@
...
+    std::vector<std::string> addresses;
+    std::vector<int> ports;
+    GetRedisShards(primary_context_->sync_context(), addresses, ports);
+    if (addresses.size() == 0 || ports.size() == 0) {
+      addresses.push_back(address);
+      ports.push_back(port);
+    }
+
+    // Populate shard_contexts.
+    for (unsigned int i = 0; i < addresses.size(); ++i) {
```
Comment (robertnishihara) [inline src/ray/gcs/client.cc]: We essentially don't use `unsigned int` anywhere in the codebase. Can you just use `int` here? This also applies below.

File: src/ray/gcs/client.cc [hunk 16]
```diff
@@ -2,51 +2,150 @@
...
+      // Slower than emplace but resource safe.
```
Comment (robertnishihara) [inline src/ray/gcs/client.cc]: I'm not sure this comment is helpful. This piece of code is definitely not performance critical. Maybe remove the comment.

File: src/ray/gcs/client.cc [hunk 17]
```diff
@@ -2,51 +2,150 @@
...
+      shard_contexts_.push_back(std::make_shared<RedisContext>());
+    }
+
+    // Call connect for all contexts. Safe to do many times.
```
Comment (robertnishihara) [inline src/ray/gcs/client.cc]: It's safe to call `Connect` multiple times? Why is that?

File: src/ray/gcs/client.cc [hunk 18]
```diff
@@ -2,51 +2,150 @@
...
+    // Here shard_contexts_.size() == addresses.size();
```
Comment (robertnishihara) [inline src/ray/gcs/client.cc]: Consider making this comment a `RAY_CHECK(shard_contexts_.size() == addresses.size())`.

File: src/ray/gcs/client.cc (hunk 8 above)
Comment (heyucongtom) [reply src/ray/gcs/client.cc]: Agreed.

File: src/ray/gcs/client.cc (hunk 15 above)
Comment (heyucongtom) [reply src/ray/gcs/client.cc]: Yep

File: src/ray/gcs/client.cc (hunk 16 above)
Comment (heyucongtom) [reply src/ray/gcs/client.cc]: Yep

File: src/ray/gcs/client.cc (hunk 18 above)
Comment (heyucongtom) [reply src/ray/gcs/client.cc]: Yep.

**Review by robertnishihara — APPROVED** [review]
